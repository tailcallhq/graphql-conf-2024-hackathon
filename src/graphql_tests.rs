use std::{fs, sync::LazyLock};

use anyhow::{anyhow, Result};
use diff_logger::DiffLogger;
use reqwest::Method;
use tracing::{error, info};

use crate::{
    query_info::Schema,
    request::{MOCK_API_CLIENT, REFERENCE_GRAPHQL_CLIENT, TESTED_GRAPHQL_CLIENT},
    type_info::Root,
};

use super::ROOT_DIR;

const NUMBER_OF_TESTS: usize = 5;

static TESTS: LazyLock<Result<Vec<String>>> = LazyLock::new(|| {
    let tests_path = format!("{ROOT_DIR}/tests");
    let mut tests = Vec::new();

    for entry in fs::read_dir(tests_path)? {
        let path = entry?.path();

        if let Some(ext) = path.extension().and_then(|s| s.to_str()) {
            if ext == "graphql" {
                tests.push(fs::read_to_string(&path)?);
            }
        }
    }

    Ok(tests)
});

pub async fn run_graphql_tests() -> Result<()> {
    info!("Run graphql assert tests");

    let tests = TESTS
        .as_ref()
        .map_err(|e| anyhow!("Failed to resolve tests due to error: {e}"))?;

    for i in 1..NUMBER_OF_TESTS {
        info!("Test iteration: {i}");

        MOCK_API_CLIENT.request(Method::POST, "reset").await?;

        for test in tests {
            let actual = TESTED_GRAPHQL_CLIENT.request(&test).await?;

            let expected = REFERENCE_GRAPHQL_CLIENT.request(&test).await?;

            let differ = DiffLogger::new();

            let difference = differ.diff(&expected, &actual);

            if !difference.is_empty() {
                error!(
                    "Actual response is not equal to expected
    Note: left is expected response -> right is actual response"
                );
                println!("{}", difference);

                return Err(anyhow!("Actual response is not equal to expected"));
            }
        }
    }

    info!("Execution of graphql tests finished");

    Ok(())
}

fn query_builder(type_name: &str) -> String {
    format!(
        r#"
        {{
          __type(name: "{}") {{
            name
            kind
            fields {{
              name
              args {{
                name
              }}
            }}
          }}
        }}
        "#,
        type_name
    )
}

pub async fn run_introspection_query() -> Result<()> {
    info!("Run graphql introspection tests");
    let query_info = include_str!("./query_info.graphql");

    // check the root query is same or not.
    let actual_value = TESTED_GRAPHQL_CLIENT.request(&query_info).await?;
    let expected_value = REFERENCE_GRAPHQL_CLIENT.request(&query_info).await?;

    let actual: Schema = serde_json::from_value(actual_value.clone())?;
    let expected: Schema = serde_json::from_value(expected_value.clone())?;
    let differ = DiffLogger::new();
    if actual != expected {
        let difference = differ.diff(
            &serde_json::to_value(actual.clone())?,
            &serde_json::to_value(expected.clone())?,
        );
        error!("Query Operation type mismatch");
        println!("{}", difference);
        return Err(anyhow!("Query Operation type mismatch"));
    }

    for (actual, expected) in actual
        .data
        .schema
        .query_type
        .fields
        .iter()
        .zip(expected.data.schema.query_type.fields.iter())
    {
        let actual_op_type = match actual.field_type.get_name() {
            Some(name) => name,
            None => continue,
        };

        let expected_op_type = match expected.field_type.get_name() {
            Some(name) => name,
            None => continue,
        };

        let actual_op_type_query = query_builder(&actual_op_type);
        let expected_op_type_query = query_builder(&expected_op_type);

        let actual: Root =
            serde_json::from_value(TESTED_GRAPHQL_CLIENT.request(&actual_op_type_query).await?)?;
        let expected: Root = serde_json::from_value(
            REFERENCE_GRAPHQL_CLIENT
                .request(&expected_op_type_query)
                .await?,
        )?;

        if actual != expected {
            let difference = differ.diff(
                &serde_json::to_value(actual)?,
                &serde_json::to_value(expected)?,
            );
            error!("Type Defination mismatch");
            println!("{}", difference);
            return Err(anyhow!("Type Defination mismatch"));
        }
    }

    info!("Execution of graphql schema validation finished");

    Ok(())
}
