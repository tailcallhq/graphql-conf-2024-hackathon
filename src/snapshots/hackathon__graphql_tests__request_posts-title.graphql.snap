---
source: src/graphql_tests.rs
expression: response
---
{
  "data": {
    "posts": [
      {
        "title": "qmevyy pdosw lazu xkbuajy qyy"
      },
      {
        "title": "mbelyf idfukxn efy rjwucow pxrw ktyes ixrcjpe"
      },
      {
        "title": "fcpbtlz fyd inryv azrz ajvrjf vqaon cmfsgy"
      },
      {
        "title": "fqlr psxoaw rbisem uxuliu tegs jpmsqoq ldc"
      },
      {
        "title": "bxx mvcy gvxzd pspedh nsyuha jcdz"
      },
      {
        "title": "ptc mtf bpkii pzqdqof hzbxw peoji"
      },
      {
        "title": "lwq qzst oaibm ziqeyq zxgn gfs mes"
      },
      {
        "title": "hgxaog leyt jkl dist fkhpume cwqusp"
      },
      {
        "title": "siqnog lcb xeomdh lilmf kgrl goz"
      },
      {
        "title": "gphmmd hfrxx ibeko pmupqhp quyknq antod qbera othhkss"
      },
      {
        "title": "bhd raj mequs ecccsdl pvbg"
      },
      {
        "title": "hfty xbixthx oupzx jcq ymrlaqd cyq pnd"
      },
      {
        "title": "gde zhwglf yar akkuyz goqirdf"
      },
      {
        "title": "vmalwtu myf oegvz jwcap phmvhx ineunqb hiqkd"
      },
      {
        "title": "rlayfe dxuyu gdieu ibqkpww uvwll qssymz ofwgs gas"
      },
      {
        "title": "euhg shjvaep wkb bqujgy eok jvg sxyc"
      },
      {
        "title": "couq bjavdl jzvsgrt vqbatqb htikt"
      },
      {
        "title": "pzcscen tni vca okiq gsjg kgjuhl"
      },
      {
        "title": "ybnp wxna fks jfq iwsc dixl ilowxz"
      },
      {
        "title": "omjvlw moghy fjc hchy znqisa arxttk"
      },
      {
        "title": "kctyra wnbc piomn lsq jfgan cnliuo xtxv"
      },
      {
        "title": "sybqszq rlewb bxf aamhwcq akff vpwn upxohz hgcle"
      },
      {
        "title": "nydgzi nwtgtx nqghk xcnqb ibtivw lwgbkti mln"
      },
      {
        "title": "ffkq nyrfp krszaql rrid yamr"
      },
      {
        "title": "uyjvxw dex zxm xslz qfb lodp ukutapi abiak"
      },
      {
        "title": "auze orhi qfu hirdb wccl cfi pnnac kskak"
      },
      {
        "title": "btmu nnznpz uugcro smynyor kryak wer kvqy"
      },
      {
        "title": "gchsgo vil xmuia mzj tswsmih"
      },
      {
        "title": "pjoc uzociq xrzgtg wqnhfli fqfgyjy chms"
      },
      {
        "title": "ufpydx dsqql piywsll qejm uuvjwyg bggd lsjdgk oazkqb"
      },
      {
        "title": "bjgtad asxq cnfv jnmlbl ezqofus ibbtn axg ifp"
      },
      {
        "title": "ntgq vovjr hjsjnh zxk bdhm hmhcrv"
      },
      {
        "title": "qvy eimkpm gggkois tlllnhv rkd dtdr"
      },
      {
        "title": "krz owlmp usath dfo aodntbj rdtcgyz"
      },
      {
        "title": "wyb uwskvij hbl otmcom lbgdmgk jriczod dfvwjr rtgh"
      },
      {
        "title": "lpwo kthww clsc demcgo xefcg kilslxj"
      },
      {
        "title": "hdixr qwwt xfnv jvalhrc rrc zjdjj"
      },
      {
        "title": "cqp svhaeb ycu uybxqd aziy icr grzcdcq"
      },
      {
        "title": "vjqwvcy ekmk ldakcrg obbsmyx firf xns pbxyckj"
      },
      {
        "title": "buva fijp qlpgvn qgfg ypw"
      },
      {
        "title": "keh lggo txrph ccecd kajhg tcfkhgf"
      },
      {
        "title": "tjbtek rgtxomr ehggyhy dho bpbwtvw wmbvxjc yowcuwf"
      },
      {
        "title": "mvqh rfne xkqkp xmq lygpwbn ckyt"
      },
      {
        "title": "qcitio zlzzzf qqdgoqe yjasa mkpc mpd pmcoi"
      },
      {
        "title": "bve mlrah rcc beh rli folkuy nsqoame"
      },
      {
        "title": "lpbiny dxs tocii auda ddrrfd"
      },
      {
        "title": "ubo zcd cjtek hrrlxm xii ogk ijz qxyaum"
      },
      {
        "title": "qzi uwccnpw anbeq mugaiw puewmgz"
      },
      {
        "title": "tzwuz tfk peijfbg ahqe yfcg shinj pukdq"
      },
      {
        "title": "clskkxr dzrzru giimyzk vxdv wqm uylk ekx orkzqcu"
      },
      {
        "title": "fezbr picssc fym gdxmc xyxv cuzvuzm"
      },
      {
        "title": "cxe pqmgb vxpypi dpnu pxditle utlao afnldlj whsixx"
      },
      {
        "title": "dnceya nrofbb bif paa gkp qtyacc cmg rzsmaw"
      },
      {
        "title": "hnii atwikpd jujc tyi stq xxjzfl gilvv jxxgkf"
      },
      {
        "title": "jomol qdv gpf upwzbsb dvsogz sdoux zfonf"
      },
      {
        "title": "wnf afbu khoapve dhiwkt mehd oxz"
      },
      {
        "title": "glbsz dkilgg ykr celcsl kkius hoj"
      },
      {
        "title": "fcvw bymrcpt iqiu ote blfln zylfei"
      },
      {
        "title": "zwsbegl owto irt lvpp fmgefvx tkk"
      },
      {
        "title": "vwei hpb rngl pjrt ahvxntc snbqxc zogod"
      },
      {
        "title": "lhdabiy ctgzo gmgtpc tmag hriai pixvs"
      },
      {
        "title": "bkldt xsee hllznwr agail caqmid utal zhfoz"
      },
      {
        "title": "qruj fkl mlwxa zoj lpt lxxac rxz"
      },
      {
        "title": "hiniu wjdtwf ygopa lfrewp rcdjdhg nli owtjv"
      },
      {
        "title": "rzeaq gdimxd pidxpz ufruopx ukvswbw xjxn"
      },
      {
        "title": "koknc mfos gueclpo iizl wtn ichhmy"
      },
      {
        "title": "nuasikv vqei rsngdsi bjo gyrybd"
      },
      {
        "title": "xop tmnwyt aia wrqpchw pxwsyvs lergnov"
      },
      {
        "title": "jxkiaji lysx jvfzgor ctpuuhv bdkk"
      },
      {
        "title": "nmlr fovnwnr gsatn hmfvgyh btmaqz krllo rfgsu ogwsy"
      },
      {
        "title": "nxrb mcaqord tyqpxh qpon rddzk ujmdh djfr"
      },
      {
        "title": "kkphv ijj uaye hwseqy lqckfxo"
      },
      {
        "title": "jyvdly pevygy czjm qclwco fsaybx yye"
      },
      {
        "title": "fqbisyv gpyml ujhoeav qujhhge olz yuu"
      },
      {
        "title": "uyf ffcdixo jgbma hwx jiwmwhw lio tmpzso mkgkg"
      },
      {
        "title": "gmgnust hodvgwg ygrc mkhpxhu igujfly kpzjut xenaxbb jovvj"
      },
      {
        "title": "kwxnxrh bqzv iorcyxf hucurk zbdx jdc ffjb vtm"
      },
      {
        "title": "fvechx qwlhw gfmk faftcgp evlse pss sra"
      },
      {
        "title": "codpxw grmlh djrdnjw yyqa utaq"
      },
      {
        "title": "wiudmx pvztzi ihmtyeh snmr apfmd"
      },
      {
        "title": "ciufdim hryrt savveoz goutn gyruuv"
      },
      {
        "title": "fbc thqow jjod fehi uaeavo jpq ddsoyy wxxjpf"
      },
      {
        "title": "tilvkb pcqhb hzrh ytsxws lfub xtoez hmbjz ilsoy"
      },
      {
        "title": "ysbbmxa ebqm emgnyxn gac gtc"
      },
      {
        "title": "tebmu qodfvz rzhdl yvkfhg iic epbihp"
      },
      {
        "title": "zwc pbka hztjog azs ivxd igglac"
      },
      {
        "title": "eer hpqwrw fou dpp sayn eblh qyy nlxhfl"
      },
      {
        "title": "mqdhhxv fgijp wjityin dfhzad emld sblx"
      },
      {
        "title": "cjjbslo zjxzv dcygbp lbs clb zwzchry rgb rcflwg"
      },
      {
        "title": "ausfzta ixtlr lfl gvb cln mtt tiwyzf shcnec"
      },
      {
        "title": "kkyqui bjbie rakjp jufa jtodtn eszg kinqy"
      }
    ]
  }
}
